builtins/amp-ad.js

```diff
@@ -244,14 +278,72 @@ export function installAd(win) {
         this.element.appendChild(this.iframe_);
 
         // Triggered by context.noContentAvailable() inside the ad iframe.
-        const unlisten = listen(this.iframe_, 'no-content', () => {
+        listenOnce(this.iframe_, 'no-content', () => {
           this.deferMutate(this.noContentHandler_.bind(this));
-          unlisten();
+        });
+        // Triggered by context.observeIntersection(…) inside the ad iframe.
+        listenOnce(this.iframe_, 'send-intersections', () => {
+          this.startSendingIntersectionChanges_();
         });
       }
       return loadPromise(this.iframe_);
     }
 
+    /** @override  */
+    viewportCallback(inViewport) {
+      // Lets the ad know that it became visible or no longer is.
+      this.sendAdIntersection_();
+      // And update the ad about its position in the viewport while
+      // it is visible.
+      if (inViewport) {
+        this.unlistenViewportChanges_ =
+            this.getViewport().onChanged(this.sendAdIntersection_.bind(this));
+      } else {
+        this.unlistenViewportChanges_();
```

Comment:
While unlikely, but let's check on null value for `this.unlistenViewportChanges_` and null it after calling it.